File: app/src/main/java/com/example/krystianwsul/organizatortest/domainmodel/dates/DayOfWeek.java
package com.example.krystianwsul.organizatortest.domainmodel.dates;

import junit.framework.Assert;

import java.text.DateFormatSymbols;
import java.util.Calendar;

public enum DayOfWeek {
    SUNDAY,
    MONDAY,
    TUESDAY,
    WEDNESDAY,
    THURSDAY,
    FRIDAY,
    SATURDAY;

    public static DayOfWeek getDayFromCalendar(Calendar calendar) {
        Assert.assertTrue(calendar != null);

        int day = calendar.get(Calendar.DAY_OF_WEEK);
        return values()[day-1];
    }

    public String toString() {
        return DateFormatSymbols.getInstance().getWeekdays()[this.ordinal() + 1];
    }

    public static DayOfWeek today() {
        return Date.today().getDayOfWeek();
    }
}

File: common/src/main/kotlin/time/DayOfWeek.kt
package com.krystianwsul.common.time

enum class DayOfWeek {
    SUNDAY,
    MONDAY,
    TUESDAY,
    WEDNESDAY,
    THURSDAY,
    FRIDAY,
    SATURDAY;

    companion object {

        private val dateCache = mutableMapOf<Date, DayOfWeek>() // because I can't hold this on a parcelable date

        fun fromDate(date: Date): DayOfWeek {
            if (!dateCache.containsKey(date)) dateCache[date] = fromDateSoy(date.toDateSoy())

            return dateCache.getValue(date)
        }

        fun fromDateSoy(dateSoy: DateSoy) = values()[dateSoy.dayOfWeekInt]

        fun fromJson(json: String) = values()[json.toInt()]

        val set by lazy { values().toSet() }
    }

    override fun toString() = format()

    fun toJson() = ordinal.toString()
}
